### PR TITLE
Expose add array types

### PR DESCRIPTION
**Pull request by thk123**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/unit/java_bytecode/java_bytecode_convert_class/add_array_types.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include <testing-utils/use_catch.h>
+#include <util/symbol_table.h>
+
+TEST_CASE("Add array types", "[core]")
+{
+  symbol_tablet symbol_table;
+  add_array_types(symbol_table);
+
+  const std::vector<std::string> array_types = {
+    "byte", "short", "int", "long", "float", "double", "boolean", "reference"};
```
Comment (jeannielynnmoulton) [inline jbmc/unit/java_bytecode/java_bytecode_convert_class/add_array_types.cpp]: char?

**Review by jeannielynnmoulton — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_class.h [hunk 2]
```diff
@@ -41,6 +41,10 @@ void mark_java_implicitly_generic_class_type(
   const irep_idt &class_name,
   symbol_tablet &symbol_table);
 
+/// Register in the \p symbol_table new symbols for the objects
+/// java::array[X] where X is byte, float, int, char...
+void add_array_types(symbol_tablet &symbol_table);
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.h]: That's quite a general name in the global namespace. Maybe `add_java_array_types`?

**Review by peterschrammel — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_class.h [hunk 3]
```diff
@@ -41,6 +41,10 @@ void mark_java_implicitly_generic_class_type(
   const irep_idt &class_name,
   symbol_tablet &symbol_table);
 
+/// Register in the \p symbol_table new symbols for the objects
+/// java::array[X] where X is byte, float, int, char...
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.h]: The complete list would be better.

Comment (allredj) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.h]: And document the `.clone` method.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.h (hunk 2 above)
Comment (allredj) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.h]: Or `add_java_primitive_array_types`?

**Review by allredj — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_class.h (hunk 2 above)
Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.h]: @allredj also adds `java::array[reference]` so that name would be inaccurate.

Comment (allredj) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.h]: Oh right. Even more reason to document properly! ;-)

**Review by smowton — APPROVED** [review]
No comments above those already raised by other reviewers

**Comment by thk123** [comment]
Thanks for the prompt reviews. All comments addressed if you care to re-review.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.h [hunk 4]
```diff
@@ -41,6 +41,13 @@ void mark_java_implicitly_generic_class_type(
   const irep_idt &class_name,
   symbol_tablet &symbol_table);
 
+/// Register in the \p symbol_table new symbols for the objects
+/// java::array[X] where X is byte, short, int, long, char, boolean, float,
+/// double and reference
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.h]: :pick: Doxygen requires full stop at the end of the brief description.

**Review by allredj — APPROVED** [review]

**Comment by thk123** [comment]
CI failure as I didn't rename the test in the makefile :cry:

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fc002bd8d521b994da5aff85f2d1e2f6ef19933a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117273894

**Pull request merged**